### PR TITLE
feat: Utility class for `Result` values

### PR DESCRIPTION
**Pull request by zharinov**
## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/util/result.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
+type Ok<T> = {
```
Comment (viceice) [inline lib/util/result.ts:1]: i would prefer interface

File: lib/util/result.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+  ok: true;
+  value: T;
+};
+
+type Err<E extends Error = Error> = {
```
Comment (viceice) [inline lib/util/result.ts:6]: same here

File: lib/util/result.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+  ok: false;
+  error: E;
+};
+
+type Res<T, E extends Error = Error> = Ok<T> | Err<E>;
```
Comment (viceice) [inline lib/util/result.ts:11]: same here

**Review by viceice — DISMISSED** [review]

**Review by rarkins — APPROVED** [review]

File: lib/util/result.ts [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+  ok: true;
+  value: T;
+}
+
+interface Err {
+  ok: false;
+  error: Error;
+}
+
+type Res<T> = Ok<T> | Err;
```
Comment (viceice) [inline lib/util/result.ts:11]: @zharinov can you please export these types in a new PR, as they are used on public API of Result class.

Comment (zharinov) [reply lib/util/result.ts:11]: What can go wrong if we don't do this? I'm quite certain these types are kind of private and I designed `Result` class to be the only API available here, but what downsides are?

**Pull request merged**